### PR TITLE
[Bug][Refactor] Fix the conflict of temp partition and dynamic partition operations

### PR DESCRIPTION
**Pull request by morningman**
The bug is described in issue: #3200.

This CL solve the problem by:
1. Refactor the alter operation conflict checking logic by introducing new classes `AlterOperations` and `AlterOpType`.
2. Allow add/drop temporary partition when dynamic partition feature is enabled.
3. Allow modifying table's property when there is temporary partition in table.
4. Make the properties `dynamic_partition.enable` optional, and default is true.

### CODE REVIEW

File: fe/src/main/java/org/apache/doris/alter/AlterOpType.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    // table property
+    MODIFY_TABLE_PROPERTY,
+    MODIFY_TABLE_PROPERTY_SYNC, // Some operations are performed synchronously, so we distinguish them by suffix _SYNC
+    // others operation, such as add/drop backend. currently we do not care about them
+    ALTER_OTHER,
+
+    INVALID_OP;
+
+    // true means 2 operations have no conflict.
+    public static Boolean[][] CONFLICT_MATRIX;
```
Comment (kangkaisen) [inline fe/src/main/java/org/apache/doris/alter/AlterOpType.java]: If true means 2 operations have no conflict. which should be `compatibilityMatrix`.

File: fe/src/main/java/org/apache/doris/alter/AlterOperations.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+    public boolean hasRollupOp() {
+        return currentOps.contains(AlterOpType.ADD_ROLLUP) || currentOps.contains(AlterOpType.DROP_ROLLUP);
+    }
+
+    public boolean hasRenameOp() {
+        return currentOps.contains(AlterOpType.RENAME);
+    }
+
+    // return true if it has specified operation
```
Comment (kangkaisen) [inline fe/src/main/java/org/apache/doris/alter/AlterOperations.java]: This comment could remove.

File: fe/src/main/java/org/apache/doris/analysis/AddPartitionClause.java [hunk 3]
```diff
@@ -46,12 +47,11 @@ public AddPartitionClause(SingleRangePartitionDesc partitionDesc,
                               DistributionDesc distributionDesc,
                               Map<String, String> properties,
                               boolean isTempPartition) {
+        super(AlterOpType.ADD_PARTITION);
         this.partitionDesc = partitionDesc;
         this.distributionDesc = distributionDesc;
         this.properties = properties;
         this.isTempPartition = isTempPartition;
-        
-        this.needTableStable = false;
```
Comment (kangkaisen) [inline fe/src/main/java/org/apache/doris/analysis/AddPartitionClause.java]: Why remove this code？

File: fe/src/main/java/org/apache/doris/alter/AlterOpType.java (hunk 1 above)
Comment (morningman) [reply fe/src/main/java/org/apache/doris/alter/AlterOpType.java]: OK

File: fe/src/main/java/org/apache/doris/alter/AlterOperations.java (hunk 2 above)
Comment (morningman) [reply fe/src/main/java/org/apache/doris/alter/AlterOperations.java]: OK

File: fe/src/main/java/org/apache/doris/analysis/AddPartitionClause.java (hunk 3 above)
Comment (morningman) [reply fe/src/main/java/org/apache/doris/analysis/AddPartitionClause.java]: Add it back

**Review by kangkaisen — APPROVED** [review]
+1. LGTM

**Pull request merged**